### DISCUSSION TITLE
Error message when adding piecewise-linear constraints

### DISCUSSION THREAD

**Discussion by lilanger** [spine-tools/SpineOpt.jl#489] [Support]
Hi all,
I tried to add a piecewise-linear constaint and get an error message because val.class_name (in SpineInterface) is not defined since it is only a Float value defining the efficiency.
_object_class_name(key, val::ObjectLike) = _object_class_name(key, val, val.class_name)
I think it is connected to the "unit_flow_coefficient" defined on the unit_to_node_user_constraint.

Maybe the class_name should be sourced from the key instead of the value? string(key)

Thanks a lot!
Lissy

Versions:
Spine Toolbox 0.6.13-dev0
[[deps.SpineInterface]]
deps = ["DataStructures", "Dates", "JSON", "LinearAlgebra", "PyCall", "Requires", "Sockets", "Statistics", "Test", "URIs"]
git-tree-sha1 = "b650639f871976722abd24be93b79e4d13cf95fc"
uuid = "0cda1612-498a-11e9-3c92-77fa82595a4f"
version = "0.8.5"

[[deps.SpineOpt]]
deps = ["Cbc", "Clp", "DataStructures", "Dates", "Distributions", "JSON", "JuMP", "LinearAlgebra", "Printf", "Requires", "SpineInterface"]
git-tree-sha1 = "35563b45fc0efc45f7f79b9e6d3426d85558e234"
uuid = "0d8fc150-4032-4b6e-9540-20efcb304861"
version = "0.6.5"

[pwl-error.txt](https://github.com/Spine-project/SpineOpt.jl/files/9647784/pwl-error.txt)

**Comment by mihlema** [comment]
Hi @lilanger,
I think maybe the problem is related to version incompatibility. spine db api has recently seen some updates which should be compatible with the latest SpineInterface (0.8.6). I just bumped this version requirement to SpineOpt master branch. Could you let me know, if upgrading SpineInterface resolves the problem for you?

**Reply by lilanger** [reply]
Unfortunately, the error remains the same.
[pwl-error_II.txt](https://github.com/Spine-project/SpineOpt.jl/files/9654689/pwl-error_II.txt)

**Reply by mihlema** [reply]
Ok thanks, one more question, the rest of SpineOpt runs fine (if you remove the user constraint)?

**Reply by lilanger** [reply]
yes, without the constraint it works fine.

**Comment by mihlema** [comment]
Just quickly following up on this discussion. The error was caused by SpineInterface struggling to write piece-wise flows. A quick fix for this situation is now available, which adds a new object class `i` to the database, and added as a dimension for unit_flow_op. However, we could think about e.g. using Maps for this purpose in the future.

**Comment by lilanger** [comment]
Hi Maren, 
thanks a lot, it splits up the curve now.
I am not sure if someone used the feature already (because the case study where it was supposed to be used in did not in the end, I think). I wanted to use it for piece-wise linear efficiencies but the functionality which is implemented so far would not be able to handle that as far as I understand.

Maybe we could discuss it on Wednesday, I think we would need SOS2-constraints and I am also not quite sure how I could add the unit_flow_coefficient (reduced efficiencies for the piece-wise linear segments) using the user constraints since the RHS cannot be a parameter (such as the unit_capacity) as far as I understand.

**Comment by mihlema** [comment]
Hey Lissy,
thanks so much for bringing this up. I am not sure if @DillonJ has used it before. Actually, we were also just talking about this. I think you are right about the SOS2-constraint. In any case, indeed let's discuss the options on Wednesday during the early user meeting, looking forward to it (unless you need an answer before Wednesday).

**Reply by lilanger** [reply]
Yes, let's do it during the meeting.

Looking forward,
Lissy

**Comment by DillonJ** [comment]
@manuelma Do you have any insight into this? I believe there are tests for piecewise linear constraints?

**Comment by nnhjy** [comment]
Using the `unit_flow_op` variable seems working well on my machine. Please try this example DB: [example.zip](https://github.com/Spine-project/SpineOpt.jl/files/9861252/example.zip). 

@lilanger Could you provide a minimal example DB which reproduce the issue you had?

**Comment by mihlema** [comment]
The problem is that the mathematical formulation of SpineOpt does not enforce the piece-wise linear operating points as intended. E.g., the only bounds on operating points are:
```
sum(unit_flow_op) = unit_flow
unit_flow_op(i) <= unit_capacity*operating_point(i)-operating_point(i-1)
```
However, this allows the model to choose operating points "as it pleases", i.e., it can operate a unit a op=4; but in reality, all previous segments first need to be activated. The current implementation will only work if efficiencies are constantly decreasing over the operating points. What we need is a piece-wise MILP representation, where operating point[i] can only be active, if all unit_flow_ops of the operating points are already operating at their maximum capacity. I've already sketched this implementation, and we can have a look at it during the user & developer calls.

**Reply by nnhjy** [reply]
Awesome, that's exactly an issue I wanted to raise on Wednesday.

**Comment by DillonJ** [comment]
Hi folks,

Sorry for arriving late into this discussion. 

The intention of this functionality was to implement piecewise linear heatrate functions where we rely on the assumption of convexity. In this case, we are assuming that, in the case of the constraint unit_pw_heat_rate, the incremental heat rates of each segment are monotonically increasing meaning we can rely on optimality for loading of the operating points in the correct order. In the case of piece-wise linear heat rates, this assumption is industry standard and relied up on in all the commercial tools that I have used. There are huge advantages to assuming monotonically increasing heat rates as we avoid the need for expensive additional integer constraints. 

The intention was never to implement a generic piecewise-linear intput / output relationship. So this functionality isn't broken, it's working as intended. However, I completely see the value of a more generic formulation that permits a more general peicewise relationship, not relying on the assumption of convexity and utilising additional integer variables, as long as the existing functionality is preserved - since we have many models relying on this. 

If we look at his another way - if someone has a heat rate function with monitonically increasing heat rates, additional integers are not necessary and they shouldn't be forced them. 

I see @nnhjy is working on a merge request. It would be great to have this functionality. I would just like to be sure that I can continue using the existing functionality without additional integer variables. @nnhjy would this be the case? I'm thinking a method parameter would be appropriate here. E.g. `unit_piecewise_method` with two options: unit_piecewise_method_enforce_ordering and unit_piecewise_method_assume_convexity.

A side point is that as a modeller and an engineer, I would always try to avoid using non-convex functions because of the performance impact. My experience is that using a piecewise linear function with an offset (the idle heat rate in the case of the unit_pw_heat_rate constraint or unit_on coefficient in the case of user cosntraints) you can still capture increasing efficiency at higher operating points. In fact, you can capture quite non convex efficiency curves this way. For example, below is a simple two-segment piecewise linear heat rate with idle heat rate of 100GJ/h and an incremental heat rate of 4 GJ/MWh from 0-200MW and 8 GJ/MWh thereafter.

![image](https://user-images.githubusercontent.com/7080191/204095878-36120f26-ec89-445c-a3e4-b7aab419b02a.png)

Thoughts?

**Reply by lilanger** [reply]
Hi Jodi,
I think the intention was never to take this out, it is great that it works without integers. Using the offset to capture increasing efficiencies also looks super interesting. I think I might not have fully grasped the power of the user constraints yet, hope I am getting there. I think we will talk about it again on Wednesday at the user meeting.

**Reply by nnhjy** [reply]
Hi Jody,
Many thanks for the elaboration and suggestions! As Lissy said, the intention was simply to extend the applicability of user constraints when dealing with segmental unit flows. The original `pw_hear_rate` constraint is independent of the "user_constraint". As such, it should still work as such within this implementation.

Here the new variable `unit_flow_op_active` that ranks unit flow segments uses the same variable type as `units_on`, not enforced to be integers by default. The problem with it being linear would be inaccuracy in the case of increasing efficiencies: low-ordered segments of a unit flow are not fully used when their successive segment gets activated.

In the case of decreasing efficiencies where low-ordered segments are born to be prioritised, there is no accuracy issue with `unit_active` being linear. The only side effect is to have some extra unnecessary linear constraints. I am not sure how badly they would hurt the model's performance. We could allow the user to exclude them by implementing the proposed `unit_piecewise_method`. Or maybe we could let performance-sensitive users go for the specific `pw_hear_rate` constraint, i.e. let the model automatically deactivate irrelevant constraints when the `pw_hear_rate` constraint is used.

**Reply by DillonJ** [reply]
Great work!

The `pw_heat_rate` constraint is not quite independent since it uses the `unit_flow_op` variables. I think it's good that the `unit_flow_op_active` variables will default to linear but still, they are not needed in the case when the heat rate function is monitonically increasing and we would still, in this case, be adding unecessary variables and constraints to the problem which has impacts everywhere (problem size, solution time, variable and constraint creation and updating result saving, reporting writing... all which can be significant in large models (the performance of which, also, we have been working very hard to improve). 

If you think of the `unit_flow_op` variables themselves, they are only created at all, if one specifies operating points. I think we shoudl stick with this principle and I feel that the unit_flow_op_active variables should only be included at all, if a user wants them (e.g. by specifying some method parameter). It's perfectly possible to do

Also, I think we also don't want to force the `unit_flow_op_active` on user_constraints either that use `unit_flow_op`, which may be monotonically increasing - and in this case also, we want the user to have the performance advantage of not needing extra variables and constraints.

To look at it another way - someone that today has a model that is working perfectly fine with monotonically increasing heat rates and/or user_constraints, if they were to merge these changes, their problem would, all of a sudden, have extra variables and constraints that are not needed (with the possibility also of unforseen impacts on the optimisation problem).
